SwiftCompilerSources/Sources/Optimizer/FunctionPasses/EffectInference.swift

```diff
@@ -0,0 +1,204 @@
+//===--- EffectInference.swift - Computes side-effects          ----------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SIL
+
+let effectInference = FunctionPass(name: "infer-effects", {
+  (function: Function, context: PassContext) in
+  var effectState = EffectState(function.entryBlock.arguments.endIndex, context.calleeAnalysis)
+
+  for block in function.blocks {
+    for inst in block.instructions {
+      effectState.update(for: inst)
+    }
+  }
+
+  context.modifyEffects(in: function, { effects in
+    effects.globalEffects = effectState.globalEffects
+    for (idx, (effect, path)) in effectState.argumentEffects.enumerated() {
+      effects.argumentEffects.append(
```

Comment:
An argument effect should only be added if
1. there is no corresponding global effect
2. It differs from the conservative default, e.g. "no read" from an in-argument, "no write" from an inout argument